tap_redshift/__init__.py

```diff
@@ -0,0 +1,421 @@
+
+        stream_version = get_stream_version(tap_stream_id, state)
+        state = singer.write_bookmark(state,
+                                      tap_stream_id,
+                                      'version',
+                                      stream_version)
+        activate_version_message = singer.ActivateVersionMessage(
+            stream=catalog_entry.stream,
+            version=stream_version
+        )
+
+        # If there's a replication key, we want to emit an ACTIVATE_VERSION
+        # message at the beginning so the records show up right away. If
+        # there's no bookmark at all for this stream, assume it's the very
+        # first replication. That is, clients have never seen rows for this
+        # stream before, so they can immediately acknowledge the present
+        # version.
+        if replication_key or bookmark_is_empty:
+            yield activate_version_message
+
+        if replication_key_value:
+            entry_schema = catalog_entry.schema
+
+            if entry_schema.properties[replication_key].format == 'date-time':
+                replication_key_value = pendulum.parse(replication_key_value)
+
+            select += ' WHERE {} >= %(replication_key_value)s ORDER BY {} ' \
+                'ASC'.format(replication_key, replication_key)
+            params['replication_key_value'] = replication_key_value
+
+        elif replication_key is not None:
+            select += ' ORDER BY {} ASC'.format(replication_key)
+
+        time_extracted = utils.now()
+        query_string = cursor.mogrify(select, params)
+        LOGGER.info('Running {}'.format(query_string))
+        cursor.execute(select, params)
+        row = cursor.fetchone()
+        rows_saved = 0
+
+        with metrics.record_counter(None) as counter:
+            counter.tags['database'] = catalog_entry.database
+            counter.tags['table'] = catalog_entry.table
+            while row:
+                counter.increment()
+                rows_saved += 1
+                record_message = row_to_record(catalog_entry,
+                                               stream_version,
+                                               row,
+                                               columns,
+                                               time_extracted)
+                yield record_message
+
+                if replication_key:
+                    state = singer.write_bookmark(state,
+                                                  tap_stream_id,
+                                                  'replication_key_value',
+                                                  record_message.record[
+                                                    replication_key])
+                if rows_saved % 1000 == 0:
+                    yield singer.StateMessage(value=copy.deepcopy(state))
+                row = cursor.fetchone()
+
+        if not replication_key:
```

Comment:
I'm not sure this logic makes sense together with https://github.com/datadotworld/tap-redshift/pull/2/files#diff-d4d5439b74cfafa4d62430a86777da3eR259

Reply:
The intention here is to activate this version at the end of the `FULL` replication since we assume that if there is no replication_key and method, we are doing a `FULL` replication and also clear the stream's version from the state.